pkg/controllers/resources/volumesnapshots/volumesnapshots/translate.go

```diff
@@ -0,0 +1,74 @@
+package volumesnapshots
+
+import (
+	volumesnapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v4/apis/volumesnapshot/v1"
+	"github.com/loft-sh/vcluster/pkg/constants"
+	"github.com/loft-sh/vcluster/pkg/util/translate"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (s *syncer) translate(targetNamespace string, vVS *volumesnapshotv1.VolumeSnapshot) (*volumesnapshotv1.VolumeSnapshot, error) {
+	target, err := s.translator.Translate(vVS)
+	if err != nil {
+		return nil, err
+	}
+	pVS := target.(*volumesnapshotv1.VolumeSnapshot)
+
+	if vVS.Annotations != nil && vVS.Annotations[constants.SkipTranslationAnnotation] == "true" {
+		pVS.Spec.Source = vVS.Spec.Source
+	} else {
+		if vVS.Spec.Source.PersistentVolumeClaimName != nil {
+			pvcName := translate.PhysicalName(*vVS.Spec.Source.PersistentVolumeClaimName, vVS.Namespace)
+			pVS.Spec.Source.PersistentVolumeClaimName = &pvcName
+		}
+		pVS.Spec.Source.VolumeSnapshotContentName = vVS.Spec.Source.VolumeSnapshotContentName
+	}
+
+	pVS.Spec.VolumeSnapshotClassName = vVS.Spec.VolumeSnapshotClassName
+	return pVS, nil
+}
+
+func (s *syncer) translateUpdate(pVS, vVS *volumesnapshotv1.VolumeSnapshot) *volumesnapshotv1.VolumeSnapshot {
+	var updated *volumesnapshotv1.VolumeSnapshot
+
+	// snapshot class can be updated
+	if pVS.Spec.VolumeSnapshotClassName != vVS.Spec.VolumeSnapshotClassName {
+		updated = newIfNil(updated, pVS)
+		updated.Spec.VolumeSnapshotClassName = vVS.Spec.VolumeSnapshotClassName
+	}
+
+	updatedAnnotations := s.translator.TranslateAnnotations(vVS, pVS)
+	if !equality.Semantic.DeepEqual(updatedAnnotations, pVS.Annotations) {
+		updated = newIfNil(updated, pVS)
+		updated.Annotations = updatedAnnotations
+	}
+
+	updatedLabels := s.translator.TranslateLabels(vVS)
+	if !equality.Semantic.DeepEqual(updatedLabels, pVS.Labels) {
+		updated = newIfNil(updated, pVS)
+		updated.Labels = updatedLabels
+	}
+
+	return updated
+}
+
+func (s *syncer) translateUpdateBackwards(pObj, vObj *volumesnapshotv1.VolumeSnapshot) *volumesnapshotv1.VolumeSnapshot {
+	var updated *volumesnapshotv1.VolumeSnapshot
+
+	// sync back the finalizers
+	if !equality.Semantic.DeepEqual(vObj.Finalizers, pObj.Finalizers) {
```

Comment:
don't sync finalizers, they could only exist in host cluster, but not in virtual cluster which could cause confusion